HW2/P5/physics.pyx

```diff
@@ -69,17 +78,37 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
+    gx = <unsigned int> (XY[i, 0] / grid_spacing)
+    gy = <unsigned int> (XY[i, 1] / grid_spacing)
+    gsize = <unsigned int> ((1.0 / grid_spacing) + 1)
+    for x in range(gx - 2, gx + 3):
+        if x < 0 or x >= gsize:
+            continue
+        for y in range(gy - 2, gy + 3):
+            if y < 0 or y >= gsize:
+                continue
+            j = Grid[x, y]
+            # Avoid double counting collisions
+            # Avoid deadlock
+            if j <= i or j > count:
+                continue
+
+            acquire(&locks[j])
+
+            XY2 = &(XY[j, 0])
+            V2 = &(V[j, 0])
+
+            if overlapping(XY1, XY2, R):
+                # SUBPROBLEM 4: Add locking
+                if not moving_apart(XY1, V1, XY2, V2):
+                    collide(XY1, V1, XY2, V2)
+
+                # give a slight impulse to help separate them
+                for dim in range(2):
+                    V2[dim] += eps * (XY2[dim] - XY1[dim])
+
+            release(&locks[j])
```

Comment:
You'd need to move the acquire within the overlapping as well. And I agree that that could positively affect your performance.